snapshot/storage/metastore.go

```diff
@@ -63,22 +67,33 @@ type transactionKey struct{}
 // TransactionContext creates a new transaction context. The writable value
 // should be set to true for transactions which are expected to mutate data.
 func (ms *MetaStore) TransactionContext(ctx context.Context, writable bool) (context.Context, Transactor, error) {
-	db, err := bolt.Open(ms.dbfile, 0600, nil)
-	if err != nil {
-		return ctx, nil, errors.Wrap(err, "failed to open database file")
+	ms.dbL.Lock()
+	if ms.db == nil {
+		db, err := bolt.Open(ms.dbfile, 0600, nil)
+		if err != nil {
+			ms.dbL.Unlock()
+			return ctx, nil, errors.Wrap(err, "failed to open database file")
+		}
+		ms.db = db
 	}
+	ms.dbL.Unlock()
 
-	tx, err := db.Begin(writable)
+	tx, err := ms.db.Begin(writable)
 	if err != nil {
 		return ctx, nil, errors.Wrap(err, "failed to start transaction")
 	}
 
-	t := &boltFileTransactor{
-		db: db,
-		tx: tx,
-	}
+	ctx = context.WithValue(ctx, transactionKey{}, tx)
 
-	ctx = context.WithValue(ctx, transactionKey{}, t)
+	return ctx, tx, nil
+}
 
-	return ctx, t, nil
+// Close closes the metastore and any underlying database connections
+func (ms *MetaStore) Close() error {
```

Comment:
When will this be called? If we don't close bolt properly, will there be corruption?

Reply:
Not called right now. Boltdb is crash safe so isn't a problem. The metadata boltdb has never been closed. We don't have a shutdown flow right now but if we add one it would be called then.

Reply:
I see. Thanks! :)